tests/Doctrine/Tests/ORM/Functional/SelfReferencingTest.php

```diff
@@ -0,0 +1,173 @@
+
+use Doctrine\ORM\Annotation as ORM;
+use Doctrine\Tests\OrmFunctionalTestCase;
+
+class SelfReferencingTest extends OrmFunctionalTestCase
+{
+    protected function setUp() : void
+    {
+        parent::setUp();
+        try {
+            $classMetadatas = [
+                $this->em->getClassMetadata(Issue7877ApplicationGenerated::class),
+                $this->em->getClassMetadata(Issue7877DatabaseGenerated::class),
+            ];
+            // We first drop the schema to avoid collision between tests
+            $this->schemaTool->dropSchema($classMetadatas);
+            $this->schemaTool->createSchema($classMetadatas);
+        } catch (Exception $e) {
+        }
+    }
+
+    public function providerDifferentEntity()
+    {
+        yield [Issue7877ApplicationGenerated::class];
+        yield [Issue7877DatabaseGenerated::class];
+    }
+
+    /**
+     * @dataProvider providerDifferentEntity
+     */
+    public function testDifferentEntity(string $class)
+    {
+        $count = count($this->sqlLoggerStack->queries);
+
+        $parent = new $class($parentId = 1);
+        $this->em->persist($parent);
+
+        $child = new $class($childId = 2);
+        $child->setParent($parent);
+        $this->em->persist($child);
+
+        $this->em->flush();
+        $this->assertCount($count + 4, $this->sqlLoggerStack->queries);
+
+        $this->em->clear();
+
+        $child = $this->em->find($class, $childId);
+        $this->assertSame($parentId, $child->getParent()->getId());
+    }
+
+    public function testSameEntityApplicationGenerated()
+    {
+        $count = count($this->sqlLoggerStack->queries);
+
+        $entity = new Issue7877ApplicationGenerated($entityId = 1);
+        $entity->setParent($entity);
+        $this->em->persist($entity);
+
+        $this->em->flush();
+        $this->assertCount($count + 3, $this->sqlLoggerStack->queries);
+
+        $this->em->clear();
+
+        $child = $this->em->find(Issue7877ApplicationGenerated::class, $entityId);
+        $this->assertSame($entityId, $child->getParent()->getId());
+    }
+
+    public function testSameEntityDatabaseGenerated()
+    {
+        $count = count($this->sqlLoggerStack->queries);
+
+        $entity = new Issue7877DatabaseGenerated();
+        $entity->setParent($entity);
+        $this->em->persist($entity);
+
+        $this->em->flush();
+        $this->assertCount($count + 4, $this->sqlLoggerStack->queries);
+        $entityId = $entity->getId();
+
+        $this->em->clear();
+
+        $child = $this->em->find(Issue7877DatabaseGenerated::class, $entityId);
+        $this->assertSame($entityId, $child->getParent()->getId());
+    }
+}
+
+/**
+ * @ORM\Entity
+ */
+class Issue7877ApplicationGenerated
+{
+    public function __construct(int $id)
+    {
+        $this->id = $id;
+    }
+
+    /**
+     * @ORM\Id
+     * @ORM\Column(type="integer")
+     * @ORM\GeneratedValue(strategy="NONE")
+     */
+    private $id;
+
+    public function getId(): ?int
```

Comment:
Use public properties, avoid any getter/setter interactions in test scope (as they add possible issues)